### ISSUE TITLE
DefaultListableBeanFactory throws a NullPointerException when resetBeanDefinition() is invoked concurrently

### ISSUE THREAD

**Issue by xuanyimo** [spring-projects/spring-framework#23542]
## Exception
![image](https://user-images.githubusercontent.com/11484269/63922943-3d9b8780-ca78-11e9-8e88-4919741b2632.png)

## Code 
```java
String beanName = ...;
BeanDefinitionRegistry registry = ((BeanDefinitionRegistry) context.getAutowireCapableBeanFactory());
        registry.removeBeanDefinition(beanName);
```

Any help is thanks!

**Comment by sbrannen** [comment]
What version of Spring Framework are you using?

And more importantly, please provide the full stack trace **as text**, not as a partial screenshot.

Thanks

**Comment by xuanyimo** [comment]
1. `Spring Boot` -> 2.1.4.RELEASE.
2. Production.
3. The other stack trace is our company code like the code above.And the full stack begin with NPE.
The parameter is beanName，we assert that is not null. And The method `removeBeanDefinition` has an assertion of it.

**Comment by xuanyimo** [comment]
```java
 xxx (1567853332284|xxx|ERROR||||||xxx:239|null
java.lang.NullPointerException: null
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resetBeanDefinition(DefaultListableBeanFactory.java:1013)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.removeBeanDefinition(DefaultListableBeanFactory.java:979)
	at xxx.shutdown(MarketQuote.java:570)
	at xxx.removeSymbol(QuoteThread.java:53)
	at xxx.handleEvent(QuoteThread.java:162)
	at xxx.run(QuoteThread.java:237))
```
Hello. I encounter this error again in production.

**Comment by xuanyimo** [comment]
I reproduced this by execute `removeBeanDefinition` concurrently.
```java
@Service
@Slf4j
public class BeanJob {

    private final ExecutorService executorService;
    private final int beanNum = 200;
    private final ApplicationContext context;
    private final BeanDefinitionRegistry registry;

    @Autowired
    public BeanJob(ExecutorService executorService, ApplicationContext context) {
        this.executorService = executorService;
        this.context = context;
        this.registry = ((BeanDefinitionRegistry) context.getAutowireCapableBeanFactory());
    }

    public static <T> T buildBean(String beanName, Class<T> clazz, List<Object> args, ApplicationContext context) {
        DefaultListableBeanFactory defaultListableBeanFactory = (DefaultListableBeanFactory) context.getAutowireCapableBeanFactory();
        BeanDefinitionBuilder beanDefinitionBuilder = BeanDefinitionBuilder.genericBeanDefinition(clazz);
        args.forEach(beanDefinitionBuilder::addConstructorArgValue);
        defaultListableBeanFactory.registerBeanDefinition(beanName, beanDefinitionBuilder.getBeanDefinition());
        return context.getBean(beanName, clazz);
    }

    @Scheduled(initialDelay = 5000L, fixedRate = 2000L)
    public void createBean() {
        for (int i = 0; i < beanNum; i++) {
            String beanName = getBeanName(i);
            buildBean(beanName, TestBean.class, Collections.singletonList(beanName), context);
            log.info("Create bean [{}]", beanName);
        }
    }

    @Scheduled(initialDelay = 5000L, fixedRate = 2000L)
    public void destroyBean() {
        for (int i = 0; i < beanNum; i++) {
            String beanName = getBeanName(i);
            executorService.execute(() -> registry.removeBeanDefinition(beanName));
            log.info("Destroy bean [{}]", beanName);
        }
    }

    private String getBeanName(int i) {
        return "TestBean" + i;
    }

    @Data
    @AllArgsConstructor
    public static class TestBean {
        private String name;
    }
}
```
I can only execute this method serially, isn't it ?

**Comment by sbrannen** [comment]
> I can only execute this method serially, isn't it ?

Yes, that appears the be the case.

`DefaultListableBeanFactory.resetBeanDefinition(String)` does not perform any synchronization on internal data structures like the `beanDefinitionMap`.

@jhoeller, is there anything you'd like to add to this discussion?

**Comment by xuanyimo** [comment]
It was okey when I generated beans concurrently. So...  Is `registerBeanDefinition` also not executed concurrently?

**Comment by sbrannen** [comment]
`DefaultListableBeanFactory.registerBeanDefinition(String, BeanDefinition)` synchronizes on the `beanDefinitionMap` if `hasBeanCreationStarted()` returns `true`. So that's likely why you didn't experience any issues registering the bean definitions concurrently.

**Comment by xuanyimo** [comment]
Okey, thank you very much!

**Comment by sbrannen** [comment]
In general, it's not recommended that one remove bean definitions concurrently, but it is also not explicitly forbidden.

So we'll see if we can provide a fix for this in `5.1.10`.

**Comment by sbrannen** [comment]
This fix will be available in the upcoming snapshots for 5.1.10 and 5.2.0.
